nixos/modules/misc/assertions.nix

```
@@ -29,6 +43,17 @@ with lib;
       '';
     };
 
+    _module.assertAndWarn = mkOption {
```

```suggestion
    assertAndWarn = mkOption {
```

The `_module` namespace should be reserved for module-internal options

---

Quoting [you](https://github.com/NixOS/nixpkgs/pull/97023#issuecomment-736023881):

> The main insight I had was that _module doesn't have a _ at the start because it's for module-internal things, but rather to avoid clashing with option names. While @roberth pointed out that we don't have a checks option at the root, these options are added to submodules as well. So introducing sugar for this could cause problems for any module or submodule which has a checks option. While there is no such option in nixpkgs, we can't be sure that users don't have one on their own.

---

That PR _was_ a module-internal implementation of assertions and warnings though, it added support for them to any module (including submodules) since `_module.checks` is added by default. This is in contrast to this PR here, which is opt-in for modules.

---

Perhaps a middle ground would be for `_module.assertAndWarn` to be a module-system-standardized _interface_ rather than a full-featured implementation of the concept?

---

I'd think it would be better then to make it a full-featured implementation. Essentially resurrect https://github.com/NixOS/nixpkgs/pull/97023 but with opt-in triggering of checks using `_module.triggerChecks` (like `_module.assertAndWarn`).

---

If we manage to trigger the checks implicitly at some point we can still deprecate `_module.triggerChecks`

---

So essentially move rename/reshape the `assertions`/`warnings` into `checks` and rename `assertAndWarn`?

Sounds good to me.

> but with opt-in triggering

So no help at all? I'm on board with that. It will be up to the user to fish the `_module.triggerChecks` out of any submodules and/or put it in front of some value within the submodule.

> If we manage to trigger the checks implicitly at some point

I can imagine helper functions and improvements to the `options` module argument to access (lists of?) submodules' `_module` value. (fyi, let's scope this out)

------

Should we add syntax sugar so users don't have to define `_module`, but rather this?

```nix
{ lib, config, ... }: {
  config = {
    # ...
  };
  checks = {  # new
    # ...
  };
}
```

---

Regarding syntactic sugar, at least for now I [still think it should not be done](https://github.com/NixOS/nixpkgs/pull/97023#issuecomment-736023881).

---

Shorthand strikes again. Fair enough.

---

I suppose users such as the NixOS toplevel could alias the `checks` as an internal `config.checks` option, at which point we're basically full circle, but at least the intent around the `_module.*` attributes will be clear?
Seems like a lot of bookkeeping just to create an illusion of a formally approved interface, but ok.

---

Another idea I just had: What if we just used https://github.com/NixOS/nixpkgs/pull/97023 as is with checking done strictly by default, but additionally enable customization of when it trigger (and whether they trigger at all). This way we can just add a couple manual fixes on top of the rare [breakages](https://github.com/NixOS/nixpkgs/pull/97023#issuecomment-748155199).

---

Oh no, I forgot to submit my huge response to this. We'll have to make do with a less detailed comment for now.

Laziness is crucial for modularity without infinite recursions. Adding strictness to a module is inherently risky, as we can not know the dependencies that user logic adds. A supposedly helpful warning must never cause evaluations to fail with an infinite recursion. Not even `config` would be a good enough trigger because the user may need a mutually recursive dependency with another submodule. For the main logic to keep working it _must not_ depend on any checks, because the checks themselves must depend on the main logic. Anything that makes creating such dependencies on checks easy would be a mistake. Instead, we could standardize on specific `_module` attributes, and leave `config` itself unencumbered by checks.

> This way we can just add a couple manual fixes on top of the rare [breakages](https://github.com/NixOS/nixpkgs/pull/97023#issuecomment-748155199).

Try to explain this design choice to someone who's confronted with a 120-item eval trace that appears to have no relation to their code, except for two items that are obviously not in a mutually recursive relationship according to any sensible logic. That's if they can even make sense of the data flow in their modules, because yes, that's a plural.

 - _just_ - no, this requires a specific skill
 - _manual fixes_ - so that will have to be us then
 - _rare_ - feel free to play with the crocodile; it rarely bites someone's head off.

Triggering must only happen "automatically" at the ultimate of top level. Anything lower than that is susceptible to infinite recursions when users indirectly create mutual dependencies between, for example, NixOS configurations in a deployment.

---

We're going to need *something* outside NixOS. It doesn't have to be perfect, and when the time comes, we deprecate the old opt-in solution.

Encountered it again today
 - https://github.com/srid/haskell-flake/issues/72#issuecomment-1423887755